src/metabase/upload.clj

```diff
@@ -533,59 +539,75 @@
-        ;; check for duplicates
-        _ (when (some #(< 1 %) (vals (frequencies normalized-header)))
-            (throw (ex-info (tru "The CSV file contains duplicate column names.")
-                            {:status-code 422})))
+        normalized-header+header (->> header
+                                      (map (fn [col]
+                                             [(normalize-column-name col) col]))
+                                      remove-auto-pk-cols)
+        normalized-header (map first normalized-header+header)
         [extra missing _both] (data/diff (set normalized-header)
-                                         (set normalized-field-names))]
-
+                                         ;; ignore the auto-pk field, it will be filled automatically
+                                         (disj (set normalized-field-names) auto-pk-column-name))]
+    ;; check for duplicates
+    (when (some #(< 1 %) (vals (frequencies normalized-header)))
+      (throw (ex-info (tru "The CSV file contains duplicate column names.")
+                      {:status-code 422})))
     (if (and (empty? extra) (empty? missing))
       {:parse-rows (fn [rows]
-                     (let [rows' rows ;; TODO: remove auto-pk columns here
+                     (let [rows' (cond->> rows
+                                   (seq auto-pk-col-indices)
+                                   (map remove-auto-pk-cols))
                            col-upload-types (map (comp base-type->upload-type :base_type fields-by-normed-name)
                                                  normalized-header)]
                        (parse-rows* col-upload-types rows')))
-       :table-col-names (map (comp :name fields-by-normed-name) normalized-header)}
+       :table-col-names (map (comp :name fields-by-normed-name) normalized-header)
+       :table-has-auto-pk? (contains? fields-by-normed-name auto-pk-column-name)}
       (throw (ex-info (tru "The schema of the CSV file does not match the schema of the table.")
                       {:status-code 422})))))
 
 (defn- append-csv!*
   [database table file]
   (with-open [reader (bom/bom-reader file)]
     (let [[header & rows] (csv/read-csv reader)
-          {:keys [parse-rows table-col-names]} (check-schema table header)]
+          {:keys [parse-rows table-col-names table-has-auto-pk?]} (check-schema table header)
+          driver (driver.u/database->driver database)
+          auto-pk-type (driver/upload-type->database-type driver ::auto-incrementing-int-pk)
+          create-auto-pk? (not table-has-auto-pk?)]
       (try
         (let [parsed-rows (parse-rows rows)]
-          (driver/insert-into! (driver.u/database->driver database)
+          (when create-auto-pk?
+            (driver/add-columns! driver
+                                 (:id database)
+                                 (table-identifier table)
+                                 {:_mb_row_id auto-pk-type}))
```

Comment:
```suggestion
                                 {:_mb_row_id (driver/upload-type->database-type driver ::auto-incrementing-int-pk)}))
```

It's only used once and is used conditionally, IMO it'll reduce the number of binding I need to keep track while reading this function.